pkg/azuredisk/azuredisk_v2.go

```diff
@@ -434,91 +455,146 @@ func (d *DriverV2) StartControllersAndDieOnExit(ctx context.Context) {
-			if err != nil || cachedAzDriverNode == nil {
-				klog.Errorf("Failed to get AzDriverNode resource from the api server. Error: %v", err)
-				time.Sleep(heartbeatFrequency)
-				cachedAzDriverNode = nil
-
-				continue
-			}
-		}
+	klog.V(1).Infof("Starting heartbeat loop with initial delay %v and frequency %v", initialDelay, heartbeatFrequency)
 
-		// Send heartbeat
-		azDriverNodeToUpdate := cachedAzDriverNode.DeepCopy()
-		timestamp := metav1.Now()
-		readyForAllocation := true
-		statusMessage := "Driver node healthy."
-		klog.V(2).Infof("Updating status for (%v)", azDriverNodeToUpdate)
-		if azDriverNodeToUpdate.Status == nil {
-			azDriverNodeToUpdate.Status = &azdiskv1beta2.AzDriverNodeStatus{}
-		}
-		azDriverNodeToUpdate.Status.ReadyForVolumeAllocation = &readyForAllocation
-		azDriverNodeToUpdate.Status.LastHeartbeatTime = &timestamp
-		azDriverNodeToUpdate.Status.StatusMessage = &statusMessage
-		klog.V(2).Infof("Sending heartbeat ReadyForVolumeAllocation=(%v) LastHeartbeatTime=(%v)", *azDriverNodeToUpdate.Status.ReadyForVolumeAllocation, *azDriverNodeToUpdate.Status.LastHeartbeatTime)
-		cachedAzDriverNode, err = azN.UpdateStatus(ctx, azDriverNodeToUpdate, metav1.UpdateOptions{})
-		if err != nil {
-			klog.Errorf("Failed to update heartbeat for AzDriverNode resource. Error: %v", err)
-			cachedAzDriverNode = nil
-		}
+	time.Sleep(initialDelay)
 
-		// If context is cancelled just return
+	ticker := time.NewTicker(heartbeatFrequency)
+	defer ticker.Stop()
+
+	for {
 		select {
+		// If context is cancelled just return
 		case <-ctx.Done():
-			klog.Errorf("Context cancelled, stopped sending heartbeat.")
+			klog.Info("Context cancelled, stopped sending heartbeat.")
 			return
-		default:
-			// sleep
-			time.Sleep(heartbeatFrequency)
-			continue
+		// Loop and update heartbeat at update frequency
+		case <-ticker.C:
+			_ = d.updateAzDriverNodeHearbeat(ctx)
+		}
+	}
+}
+
+// registerAzDriverNode initializes the AzDriverNode object
+func (d *DriverV2) registerAzDriverNode(ctx context.Context) error {
+	innerCtx, w := workflow.New(ctx, workflow.WithDetails(consts.NamespaceLabel, d.objectNamespace, consts.NodeNameLabel, d.NodeID))
+
+	err := d.updateOrCreateAzDriverNode(innerCtx, azDriverNodeInitializing)
+
+	w.Finish(err)
+
+	return err
+}
+
+// updateAzDriverNodeHearbeat updates the AzDriverNode health status
+func (d *DriverV2) updateAzDriverNodeHearbeat(ctx context.Context) error {
+	innerCtx, w := workflow.New(ctx, workflow.WithDetails(consts.NamespaceLabel, d.objectNamespace, consts.NodeNameLabel, d.NodeID))
+
+	err := d.updateOrCreateAzDriverNode(innerCtx, azDriverNodeHealthy)
+
+	w.Finish(err)
+
+	return err
+}
+
+// updateOrCreateAzDriverNode updates the AzDriverNode status, creating a new object if one does not exist
+func (d *DriverV2) updateOrCreateAzDriverNode(ctx context.Context, status azDriverNodeStatus) error {
+	logger := logr.FromContextOrDiscard(ctx)
```

Comment:
Is the intent here to use a plain logger for the heartbeat update loop and context logger for when registering `AzDriverNode`?

Reply:
In both cases, this will the workflow logger. Long term, I'd like us to move to a model where we always get the logger from the context rather than assume klog.

Reply:
It seems like the context fed in this loop doesn't contain worklow though. Am I missing something?
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/4e6c6da101faf1ec382552ef3ee57b427b925b3a/pkg/azuredisk/azuredisk_v2.go#L522-L524

Reply:
I agree that we should move to getting logger from context long term.

Reply:
`updateAzDriverNodeHeartbeat` creates a workflow.

Reply:
It does but I don't think the context in `updateAzDriverNodeHeartbeat` is being passed onto `runAzDriverNodeHeartbeatLoop` though.

Reply:
We sync'ed up offline - yes, the main heartbeat loop will use the plain logger for now. Longer term, we should add `logr`-based loggers to the context passed here and use it.